src/OrbitBase/include/OrbitBase/ParameterPackTrait.h

```diff
@@ -0,0 +1,203 @@
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_PARAMETER_PACK_TRAIT_H_
+#define ORBIT_BASE_PARAMETER_PACK_TRAIT_H_
+
+#include <stddef.h>
+
+#include <tuple>
+#include <type_traits>
+
+namespace orbit_base {
+template <template <typename...> typename, typename...>
+struct ParameterPackTrait;
+}
+
+namespace orbit_base_internal {
+
+// This is a helper type that is needed to implement `ParameterPackTrait::RemoveDuplicateTypes()`.
+// It is based on recursion - AFAIK there is currently no other way to implement this.
+// The recursion depth is O(N), the number of type comparisons is O(N^2). This is far from great,
+// but since our use cases are usually N \in {1, 2, 3}, it's also not that concerning.
+template <typename Output, typename Input>
+struct DeduplicatedTuple;
+
+template <typename... OutputTypes, typename FirstInput, typename... OtherInputTypes>
+struct DeduplicatedTuple<std::tuple<OutputTypes...>, std::tuple<FirstInput, OtherInputTypes...>> {
+  using type = typename std::conditional_t<
+      orbit_base::ParameterPackTrait<std::tuple, OutputTypes...>::template Contains<FirstInput>(),
+      typename DeduplicatedTuple<std::tuple<OutputTypes...>, std::tuple<OtherInputTypes...>>::type,
+      typename DeduplicatedTuple<std::tuple<OutputTypes..., FirstInput>,
+                                 std::tuple<OtherInputTypes...>>::type>;
+};
+
+template <typename... OutputTypes>
+struct DeduplicatedTuple<std::tuple<OutputTypes...>, std::tuple<>> {
+  using type = std::tuple<OutputTypes...>;
+};
+
+template <typename Tuple>
+using DeduplicatedTuple_t = typename DeduplicatedTuple<std::tuple<>, Tuple>::type;
+
+template <typename Tuple>
+struct TupleToParameterPackTraits;
+
+template <typename... Types>
+struct TupleToParameterPackTraits<std::tuple<Types...>> {
+  using type = orbit_base::ParameterPackTrait<std::tuple, Types...>;
+};
+
+template <typename Tuple>
+using TupleToParameterPackTraits_t = typename TupleToParameterPackTraits<Tuple>::type;
+
+}  // namespace orbit_base_internal
+
+namespace orbit_base {
+
+// This is a trait that helps you deal with a parameter pack `Types...`. It also captures a variadic
+// holder template type (any template that takes a parameter pack as its argument - like
+// `std::variant` or `std::tuple`). For example `ParameterPackTrait<std::tuple, int, char>`
+// describes the type `std::tuple<int, char>`.
+//
+// An instance of this trait can be constructed either through default construction or from an
+// existing value using the `MakeParameterPackTrait` constexpr factory function below. Look there
+// for more details.
+//
+// The trait provides a variety of helper functions to deal with parameter packs. They are
+// documented below in the code.
+template <template <typename...> typename VariadicHolder, typename... Types>
+struct ParameterPackTrait {
+  // Returns true iff the given type `T` is part of the parameter pack.
+  template <typename T>
+  [[nodiscard]] constexpr static bool Contains() {
```

Comment:
Yes, you're totally right - a lot of stuff here could be constants.

But I didn't make them constants on purpose. I wanted this to feel like normal code that you would also write for runtime execution but constexpr-ed. I have even considered making these functions `const` instead of `static`. I'm worried that making them constants makes the code harder to understand for folks that are less familiar with TMP. I think functions feel more natural and have no downside. They also allow for more complex logic with early returns for example. Granted that's not needed here at the moment.

In my next PR, I have this function for example:

```cpp
  template <typename... Types>
  constexpr static bool CanBeConstructedFromTypesAndIsNotCopy() {
    constexpr ParameterPackTrait<AnyErrorOf, ErrorTypes...> kThisTrait{};
    constexpr ParameterPackTrait<AnyErrorOf, Types...> kOtherTrait{};

    // We return false if `AnyErrorOf<ErrorTypes...>` is the same type as `AnyErrorOf<Types...>`.
    // This avoids collisions with the copy and move constructors/assignment operators.
    if (kThisTrait == kOtherTrait) return false;

    // We only allow conversion from an instance of `AnyErrorOf<Types...>` with `Types` being a
    // subset of `ErrorTypes`.
    return kThisTrait.IsSubset(kOtherTrait);
  }
```

This feels much more natural to me, even though it's technical a constant. And it's all constexpr and I'm using it as a predicate for SFINAE.

Technically also the constexpr comparison operators that I declare below are not needed. It's just a nice way to write what `std::is_same_v` can also do.

WDYT? Do you see any disadvantages of functions over constants?

Reply:
Why do we have numerous integer constants like `kInvalid` and not functions? 

From a function call-site, it's absolutely not clear whether it's a compile-time computation or not.

Also, I've not suggested to change `IsSubset(const ParameterPackTrait<VariadicHolder, Others...>&)`. I presumed, you don't want to write `kThisTrait.IsSubset<decltype(kOtherTrait)>` too often. Here I can see how functions help to improve readability. But there is a downside --  does the result depends on the value (and not just the type) passed to the function?

Reply:
BTW. If this stuff will be used in SFINAE, how many "folks that are less familiar with TMP" will be reading it anyway? 

... and what's about compile-time (yeah, not the strongest of all arguments)? Will compiler "understand", it shouldn't re-calculate the function twice if it's called with the same parameters?

Reply:
Changed.